택배 송장 제조기 [STEP1] gogochang

gogochang opened the pull request:
@GREENOVER 
안녕하세요. gogochang입니다.
택배 송장 제조기 [STEP1] 미션 요구사항 PR입니다.
STEP1을 진행하며 고민과 과정을 작성했습니다.
잘못 이해한 부분이나 더 좋은 방법에 대해 조언과 방향을 지시해주시면 감사하겠습니다.

## 고민했던 점
### 객체미용체조 7원칙 ‘2개 이상의 원시타입 프로퍼티를 갖는 타입 금지

#### 새로운 타입으로 분리
- **ParcelInfomation** 타입은 여러가지 원시타입 프로퍼티를 가지고 있었습니다.
주소, 보내는사람이름, 보내는사람 정보, 비용, 할인, Discount 타입에 따른 할인된비용 에서
보내는사람의 이름과 정보 프로퍼티를 **ReciverInfo** 타입으로,
비용,할인 과 같은 금액과 관련된 프로퍼티를   **ParcelCost** 타입으로 새로 생성해 주었습니다.
주소, ReciverInfo, ParcelCost를 다시 하나로 묶어 **ParcelInfo**라는 새로운 타입을 생성하여 분리하였습니다.

#### 매직넘버
- **Discount** 열거형의 케이스에 따라 할인율이 달랐습니다. 이 할인율을 계산하는 코드가 "5x4", "2"라는 매직넘버를 사용하고 있었습니다.
**객체미용 3원칙 원시값과 문자열의 포장**을 적용하기 위해 **DiscountRate** 열거형 생성하여 0.8, 0.5로 Double 타입프로퍼티로 정의해 주고
**discountedCost** 계산을 곱셈으로 변경하여 안전성과 가독성을 개선했습니다.

#### 4원칙 한 줄에 한 점만 사용
- 위 수정사항을 적용 후 **InvoidView** 타입에서 **ParcelInfomation**의 정보를 가져오는 부분에서 여러 점을 사용하게 되어 결합도가 증가하는 문제가 생겼습니다. 
위 원시타입을 분리하는 과정에서 만든 **ParcelInfo** 에서 프로퍼티값을 반환하는 메서드를 생성했습니다.
그리고 이 ParceInfo를 가지는 ParceInfomation에서 ParceInfo의 메서드를 사용하여
각 데이터를 반환하는 메서드를 추가하여 현재 타입이 가지고있는 인스턴스를 통해서만 가져올 수 있도록 수정하여
여러 점을 사용하는 부분을 해결했습니다.

### SOLID SRP 적용
#### 책임 분리
- **ParcelOrderProcessor** 타입의 이름에서도 알 수 있듯이 소포 주문을 처리하는 기능을 담당하고 있습니다. SRP는 단일 책임 원칙(Single Responsibility Principle)을 따르기 위해서는 소포 주문과 관련없은 데이터 베이스 저장기능은 다른 타입으로 구분하여 하나의 책임만 가지도록 수정해야 합니다.
요구사항에 따라 **DatabaseParcelInformationPersistence** 타입을 생성하고 데이터 베이저장 메서드를 옮긴 후 ParcelOrderProcessor 타입의 내부에 프로퍼티로 만들어 저장을 요청하도록 수정했습니다.

#### 어떻게 적용할까?
- 앞으로 코드를 작성하며 타입을 생성할 때 강의에 나와있는 주요 체크포인트를 참고하고 "과연 이 타입이 하나의 책임을 가지고 있는가?"를 끊임 없이 확인해야겠습니다. 한 가지 책임만 가질 수 있도록 하여 가독성, 유지보수에 좋은 코드를 작성할 수 있도록 신경써야겠습니다.

### SOLID DIP 적용
#### 추상화프로토콜에 의존하도록 수정
- SOLID SRP까지 적용하고 나면,  **ParcelOrderProcessor**(상위 계층 모듈)이 **DatabaseParcelInformationPersistence**(하위 계층 모듈)의 세부사항에 의존하고 있는 문제가 발생하여 DIP(Dependency Inversion Principle)원칙에 위배됩니다. 
또 databaseParcelInformationPersistence를 변경하려면 내부 코드를 다시 수정하고 OCP(Open Closed Principle)에도 위배되는 문제가 발생합니다. 이러한 문제를 해결하기 위해 프로토콜을 사용한 추상화한 **ParcelInformationPersistence** 프로토콜을 사용하여 하위 계층 모듈의 세부사항이 아닌 추상화를 소유하도록 했습니다.

#### 어떻게 적용할까?
- 추상화된 프로토콜을 사용하여 기존 코드는 수정하지 않고 타입 내부의 구체 타입 변경이 용이하며,
서로의 의존성을 낮추고 보다 유연한 코드를 작성할 수 있다고 생각했습니다. 타입의 재사용성을 고민하고 공통부분이 무엇인지 생각하여 추상화할 수 있는 연습을 해야겠습니다.

### 파일 생성
- **ParcelProcessor**는 소포를 처리하는 이름을 가지고 있어 **ParcelInfomation**라는 이름의 Swift파일을 따로 생성했습니다.
- 소포를 처리하기 위한 데이터를 정리했습니다.

---

## 해결이 되지 않은 점
### 파일 분리
- 하나의 Swift파일 내부에 서로 다른 클래스,구조체,열거형이 같이 있는것을 정리하고 싶었습니다.
서로 관계가 있다고 생각하여 하나의 Swift파일에 작성했지만 이 기준을 정확히 잡지못했습니다.

---

## 조언을 얻고 싶은 부분
### 4원칙 한 줄에 한 점만 사용
- **InvoiceView**의 **ParcelInfomation의 데이터**를 가져오는 부분에서 건너 건너 접근하며 결합도가 높아진 문제가 있었습니다. 이를 해결하기 위해 Struct가 가지고있는 하위 Struct의 프로퍼티를 가져오기위한 메서드를 추가한 저의 방법이 문제없는지, 더 좋은 방법이 있는지 궁금합니다.

### SOLID DIP 적용
- "실제 구체타입은 추상화를 바라보고있고 상위 계층 모듈은 하위 계층 모듈의 추상화된 프로토콜을 소유하는 것" 그래서 "서로 의존하지 않고 추상화 된 프로토콜을 의존하고 있는 것" 로 이해했습니다. 코드상에서 미션 요구사항을 적용할 때 **ParcelInformation**를 하위 계층 모듈로 둘 때 실제 세부사항에 직접 소유하고 있는 부분은 모두 추상화된 **ParcelInformationProvider** 프로토콜로 수정했습니다. 이 부분에서 제가 이해한것과 적용한것이 맞는지 확인하고 싶습니다.

### 내가 잘못 이해하진 않았는가?
- 제가 고민했던 부분과 잘못이해한 부분이 있어 조언해주시면 감사하겠습니다.

gogochang committed:
택배 송장 제조기 [STEP1] gogochang
    - 객체미용체조 7원칙 '2개 이상의 원시타입 프로퍼티를 갖는 타입 금지'
    - SOLID SRP 적용
    - SOLID DIP 적용

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelInfomation.swift:
```diff
@@ -0,0 +1,110 @@
+    func getDiscountedCost() -> Int
+}
+
+class ParcelInformation: ParcelInformationProvider {
```
class로 타입을 지정하신 이유가 궁금합니다 🙋🏻

gogochang replied:
@GREENOVER 
솔직히 말하면... 기존 코드에 작성한 그대로 사용했습니다.
값타입, 참조타입 이정도의 차이점만 알고 있었고 
언제 사용해야하는지 그린님 질문에 고민이 되었습니다.

그래서 찾아봤습니다
* **class**
참조타입이기 때문에 실제 고유한 값을 구별할때,
다른 클래스를 상속받을 때
objective-c와 상호작용할 때 (클래스 상속때문에)

* **struct**
swift에서 struct는 클래스의 기본 기능이 사용가능합니다. 
그래서 디폴트로 사용되는데 (권장?)
위 코드에서는 class만 가능한 다른 class를 상속받을 필요는 없어보입니다.
또 스택영역에 할당되기 때문에 struct를 사용할 수 있는 상황이라면
struct를 사용하는게 바람직하다고 생각해서 **struct**로 변경해야될것같아요

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelInfomation.swift:
```diff
@@ -0,0 +1,110 @@
+struct ParcelCost {
+    let deliveryCost: Double
+    private let discount: Discount
+    var discountedCost: Int {
+        switch discount {
+        case .none:
+            return Int(deliveryCost)
+        case .vip:
+            return Int(deliveryCost * DiscountRate.vip)
+        case .coupon:
+            return Int(deliveryCost * DiscountRate.coupon)
+        }
+    }
```
지금도 충분히 괜찮지만 SRP를 한번 더 생각해서 극한으로 하나의 책임만 갖도록 해본다면 현재 ParcelCost는 할인 계산 로직까지 들어가 있어서 이를 분리해볼 수도 있을것 같습니다 🙋🏻

해당하는 프로토콜로 추상화 및 채택하여 가져가볼 수 도 있을것 같은데요.

말로 헷갈릴 수 있어 제가 생각한 흐름을 간단히 코드로 나타내볼께요..!
```swift
protocol DiscountMethod {
  func calculateDiscount(cost: Double) -> Int
}

struct VIPDiscountMethod: DiscountMethod {
  func calculateDiscount(cost: Double) -> Int { 
    return Int(cost * DiscountRate.vip)
  }
}

struct ParcelCost {
  ...
  let discountMethod: DiscountMethod
  var discountedCost: Int {
    discountMethod.calculateDiscount(cost: deliveryCost)
  }
  ...
}
```
요런 방식으로 역할을 좀 더 분리할 수 있을것 같습니다 🙋🏻
(네이밍은 신경쓰지 않아서 누추하지만 흐름만 봐주시면 감사하겠습니다ㅎ..)

GREENOVER replied:
추가로 추상화를 통해 현재 ParcelCost 구조체에서 Discount 열거 타입을 직접 의존하고 있는 형태도 같이 개선한다면 DIP도 좀 더 고려되지 않을까 생각듭니다~!

GREENOVER replied:
앗...! 요게 스텝 2에 해당하는 비슷한 내용이였네요~!
비슷한 지점이 많아 스텝 2에서 적용해보셔도 좋을것 같습니다 🙋🏻

gogochang replied:
@GREENOVER 
* 할인 계산하는 로직을 분리해 주어서 SRP
* DiscountMethod 추상화된 프로토콜을 ParcelCost 에서 프로퍼티로 소유하고 있는 형태를 DIP

추가로,
DiscountMethod로 확장을하여 새로운 DiscountMethod 프로토콜을 채택하여 새로운CouponDiscountMethod를 생성한다면 확장에는 열려있고, ParcelCost 내부의 코드는 변경에 닫혀있기 때문에 OCP의 원칙도 따르는게 맞을까요??
STEP2에 적용해보겠습니다!

GREENOVER replied:
네네 맞습니다~!
스텝2에서 한번에 같이 해보시죠ㅎㅎ

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelInfomation.swift:
```diff
@@ -0,0 +1,110 @@
+    
+    // 객체 미용 체조, 4원칙 한 줄에 한 점만 사용
+    // 주소를 반환합니다.
+    func getAdress() -> String {
```
번외로 오탈자도 수정되면 좋을것 같습니다 🙋🏻

gogochang replied:
넵

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelInfomation.swift:
```diff
@@ -0,0 +1,110 @@
+}
+
+// 할인 타입
+enum Discount: Int {
+    case none = 0, vip, coupon
+}
```
위 코멘트의 흐름으로 프로토콜로 분리한다면 Discount 열거 타입을 추후에는 없애도 될것 같습니다!
물론, 해당 ParcelInformation 인스턴스를 생성하는 부분도 같이 수정은 되어야 하겠지만요 😃

gogochang replied:
@GREENOVER 
네, DiscountMethod로 새로운 타입을 확장해서 사용하면 되기 때문에 Discount열거형으로 타입을 나눌필요가 없어보여요.
맞나요?

gogochang replied:
이거도 STEP2에 같이 적용하겠습니다

GREENOVER replied:
네 이해하신것이 맞습니다ㅎㅎ

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelOrderView.swift:
```diff
@@ -70,17 +70,19 @@ class ParcelOrderView: UIView {
-                                                         deliveryCost: cost,
-                                                         discount: discount)
+        // 객체미용체조 7원칙 '2개 이상의 원시타입 프로퍼티를 갖는 타입 금지'를 적용하면서 변경
+        let parcelInformation: ParcelInformationProvider = 
+        ParcelInformation(parcelInfo: ParcelInfo(address: address,
+                                                 receiver: ReceiverInfo(receiverName: name,
+                                                                        receiverMobile: mobile),
+                                                 cost: ParcelCost(deliveryCost: cost,
+                                                                  discount: discount)))
```
리팩토링과는 조금 논외인 코드 컨벤션 관련 의견이지만, 쪼금 가독성이 떨어지는 느낌인데 가독성을 조금 더 올려볼 수 있을까요..!?
코드 컨벤션이라 정말 개개인 스타일의 차이겠지만, 저는 주로 파라미터 개행에 대해 이런 스타일을 사용하며 가볍게 봐주셔도 좋을것 같습니다ㅎㅎ
```swift
let parcelInformation: ParcelInformationProvider = ParcelInformation(
  parcelInfo: ParcelInfo(
    address: address,
    receiver: ReceiverInfo(
      receiverName: name,
      receiverMobile: mobile
    ),
    cost: ParcelCost(
      deliveryCost: cost,
      discount: discount
    )
  )
)
```

gogochang replied:
@GREENOVER 
그린님의 파라미터 개행 스타일이 저도 마음에 듭니다. 따르겠습니다.

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelProcessor.swift:
```diff
@@ -6,49 +6,32 @@
-
 class ParcelOrderProcessor {
+    // SOLID DIP 적용 추상화된 프로토콜을 가지도록 수정
+    let databaseParcelInformationPersistence: ParcelInformationPersistence
```
private 해도 되지 않을까요..!? 🤔

gogochang replied:
넵

GREENOVER reviewed (approved):
@gogochang 
안녕하세요 고고창..!
고고창이라고 부르면 되는거겠죠?ㅎㅎ 
저는 편하게 그린이라고 불러주세요 🙋🏻
우선 디코로도 가볍게 인사를 나눴지만 첫 PR이라 한번 더 인사드려요!
3주간 잘부탁드립니다 🙇🏻
스텝1 수행하시느라 너무 수고 많으셨습니다 👏👏👏
요구사항들을 꼼꼼히 잘 수행하시면서 군더더기 없이 스텝1에 맞게 리팩토링 되었다고 생각해요.
그렇기에 저또한 고민하셨던 포인트들과 리팩토링 한 흐름에 대해 친절히 PR 메시지로 잘 담아주셔서 잘 이해할 수 있었고, 전체적인 스텝1 방향에서 추가적으로 크게 더 리팩토링 되어야할 만한 사항들은 없다고 생각합니다.
그럼으로 한단계 한번 더 다른 시야에서 바라보면서 몇가지 코멘트를 남겨봤습니다 😀
***
### 해결되지 않으신 부분에 대한 저의 의견
- 파일 분리
  - 정말 아시겠지만, 파일 및 그룹핑과 같은 부분들을 나누는것도 개개인의 스타일의 차이이겠지만, 저라면 이렇게 했을것 같다라는 하나의 의견을 드려봅니다ㅎㅎ
  저는 보통 하나의 객체에 대해 최대한 하나의 파일로 가져가고 분리하는 편입니다.
  그렇기에, ParcelInformation 모델 파일 코드로 예시를 들어보면, 하위 Parcelnfo, ReceiverInfo들을 각각의 파일로 나누고 해당 관련있는 파일들끼리 하나의 그룹으로 묶는 스타일을 가져가려합니다.
  또한, ParcelOrderProcessor와 DatabaseParcelInformationPersistence 객체도 별도 파일로 분리할 것 같아요!
  추후 생각해본다면, DatabaseParcelInformationPersistence는 ParcelOrderProcessor에서만 사용되는것이 아닌 더 확장성이 있을 수 있다고 가정한다면 역할이 더 크고 관심사가 다르기에 별도로 분리하는것이 더 적절하지 않을까 생각합니다 🙋🏻
***
### 궁금하신 부분에 대한 저의 의견
- 4원칙 한 줄에 한 점만 사용
  - 저는 현재 구현하신 방식에 동의합니다!
  물론, 메서드를 사용하지 않고 연산 프로퍼티를 활용하거나 등 방법은 더 있겠지만, 지금과 같이 메서드를 활용하여 직접 프로퍼티에 접근하지 않는 방식 자체가 더 유지보수 측면에서 편리하고 결합도도 낮춰줄 수 있다고 생각합니다 😃
- SOLID DIP 적용
  - 네 이 부분도 스텝 1 DIP 요구사항에 맞춰 적절히 리팩토링 해주셨기에 개선을 요하는 포인트는 현재 없다고 생각합니다!
- 내가 잘못 이해하진 않았는가?
  - 전체 코드를 리뷰했을때도 스텝 1의 요구사항을 놓치지 않고 잘 수행해주신걸로 보여요.
  그렇기에 현재 코드에서는 잘못 이해하고 놓치거나 이상한 포인트는 찾을 수 없었습니다!
***
다시 한번 스텝 1 고생 많으셨습니다ㅎㅎ
요구사항을 충분히 잘 만족하셔서 스텝 1 어프로브하겠습니다.
혹시 수정하실 부분이 있으실지 최종 확인을 위해 머지는 하지 않겠습니다~!
언제든지 확인해보시면서 머지 요청을 해주세요..!
(디엠으로 머지 요청 주셔도 좋습니다~)

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelInfomation.swift:
```diff
@@ -0,0 +1,110 @@
+}
+
+// 보내는사람 정보
+struct ReceiverInfo {
+    let receiverName: String
+    let receiverMobile: String
+}
```
하위 구조체에서도 프로퍼티에 직접 접근보다 메서드를 이용해 접근하는건 어떨까요?
해당 프로퍼티를 현재는 직접 접근할 수 있으니 데이터가 안전하지 않을 수 있을것 같아서요~!
private (set)해도 될것 같구요 🤔
다른 하위 Struct들도 같이 고려해보면 좋을것 같습니다~!

gogochang replied:
@GREENOVER 
하위 구조체에도 메서드를 이용해 접근할 수 있도록 수정해보겠습니다.

private (set)을 사용하면 mutating을 사용해야 하겠지만 
그래도 변경 가능성의 여지가 있다고 생각해서
메서드를 이용해 접근해보려고하는데 그린님의 의견은 어떠신가요??

GREENOVER replied:
네네 메서드로 이용해서 접근하는것도 충분히 괜찮다고 생각합니다~!
그렇다면 추가로 해당 값들을 메서드로만 접근하고 직접 접근은 하지 못하도록 은닉해도 좋을것 같아요!

gogochang committed:
택배 송장 제조기 [STEP1] gogochang
	- class -> struct 수정
	- adress -> address 오탈자 수정
	- private 변경
	- 파라미터 개행스타일로 변경
	- 하위 구조체 접근 메서드 방식 추가
